### PR TITLE
Shovel Silo Loader Mode

### PR DESCRIPTION
**Pull request by schwiti6190**
Functions:
- Loading and Unloading with a wheel loader or similar vehicles.
- Loading from a bunker silo and from heaps on fields.
- Unloading to trailers and unload triggers.
- Automatic shovel position control( Loading, Transport, Pre Unload, Unload ).
- Trailer unload searches for trailer nearby and unloads to them, similar to combine self unload.
- Unload triggers can be selected on the AI menu map.
- Added 4x Map AI Map zoom
- Quality of life improvements for vehicle config

### CODE REVIEW

File: scripts/gui/CoursePlot.lua [hunk 1]
```diff
@@ -79,44 +79,11 @@ function CoursePlot:setStopPosition( x, z )
...
-
-	return objectX, objectY, rot, true
-end
-
-function CoursePlot:screenToWorld( x, y )
-	local worldX = ((x - self.x) / self.scaleX) - self.worldOffsetX
-	local worldZ = ((y - self.y - self.height) / -self.scaleZ) - self.worldOffsetZ
-	return worldX, worldZ
-end
+-- function CoursePlot:screenToWorld( x, y )
```
Comment (pvaiko) [inline scripts/gui/CoursePlot.lua:82]: dead code

File: scripts/gui/CpStatus.lua [hunk 2]
```diff
@@ -132,11 +132,16 @@ function CpStatus:getCompactionText(withoutPercentageSymbol)
 end
 
 function CpStatus:getSiloFillLevelPercentageLeftOver(withoutPercentageSymbol)
-    if self.isActive and self.fillLevelPercentageLeftOver ~=nil  then 
-        if withoutPercentageSymbol then 
-            return tostring(self.fillLevelPercentageLeftOver)
-        end
-        return string.format('%d%%', self.fillLevelPercentageLeftOver)
+    -- if self.isActive and self.fillLevelPercentageLeftOver ~=nil  then
```
Comment (pvaiko) [inline scripts/gui/CpStatus.lua:135]: dead code

File: scripts/pathfinder/PathfinderUtil.lua [hunk 3]
```diff
@@ -895,13 +914,17 @@ end
...
----@param offFieldPenalty number penalty to apply to nodes off the field
----@param areaToAvoid PathfinderUtil.NodeArea nodes in this area will be penalized so the path will most likely avoid it
----@param areaToIgnoreFruit PathfinderUtil.Area area to ignore fruit
+---@param vehiclesToIgnore table[]|nil list of vehicles to ignore for the collision detection (optional)
+---@param maxFruitPercent number|nil maximum percentage of fruit present before a node is marked as invalid (optional)
+---@param offFieldPenalty number|nil penalty to apply to nodes off the field
+---@param areaToAvoid PathfinderUtil.NodeArea|nil nodes in this area will be penalized so the path will most likely avoid it
+---@param areaToIgnoreFruit PathfinderUtil.Area|nil area to ignore fruit
+---@return PathfinderInterface pathfinder
+---@return boolean done finished pathfinding?
```
Comment (pvaiko) [inline scripts/pathfinder/PathfinderUtil.lua:925]: PathfinderInterface:resume(...) explains what these mean, we use coroutines here to be able to give the control back to the game periodically while the pathfinder is running.

File: scripts/specializations/CpAIWorker.lua [hunk 4]
```diff
@@ -484,16 +495,86 @@ function CpAIWorker:stopCpDriver()
...
 	self:raiseAIEvent("onAIFieldWorkerEnd", "onAIImplementEnd")
-
 end
 
 function CpAIWorker:getCpDriveStrategy()
     local spec = self.spec_cpAIWorker
     return spec.driveStrategy
 end
 
+function CpAIWorker:getCpReverseDrivingDirectionNode()
```
Comment (pvaiko) [inline scripts/specializations/CpAIWorker.lua:506]: Is this even used anywhere?

File: scripts/specializations/CpAIWorker.lua [hunk 5]
```diff
@@ -484,16 +495,86 @@ function CpAIWorker:stopCpDriver()
...
+        spec.reverseDrivingDirectionNode =
+            CpUtil.createNewLinkedNode(self, "realReverseDrivingDirectionNode", self:getAIDirectionNode())
+        setRotation(spec.reverseDrivingDirectionNode, 0, math.pi, 0)
+    end
+    return spec.reverseDrivingDirectionNode
+end
+
+--- Fixes the ai reverse node rotation for articulated axis vehicles,
+--- if the node is pointing backwards and not forwards.
+function CpAIWorker:getAIReverserNode(superFunc)
```
Comment (pvaiko) [inline scripts/specializations/CpAIWorker.lua:518]: doesn't this make AIUtil.getArticulatedAxisVehicleReverserNode and its usage obsolete?

File: scripts/specializations/CpAIWorker.lua [hunk 6]
```diff
@@ -484,16 +495,86 @@ function CpAIWorker:stopCpDriver()
...
+    local spec = self.spec_cpAIWorker
+ --   if not self:getIsCpActive() then return superFunc(self) end
+    if self.spec_articulatedAxis and self.spec_articulatedAxis.aiRevereserNode then
```
Comment (pvaiko) [inline scripts/specializations/CpAIWorker.lua:521]: do they use the mistyped version "revereser" everywhere? never the correct one?

File: scripts/specializations/CpAIWorker.lua [hunk 7]
```diff
@@ -484,16 +495,86 @@ function CpAIWorker:stopCpDriver()
...
+        setRotation(spec.reverseDrivingDirectionNode, 0, math.pi, 0)
+    end
+    return spec.reverseDrivingDirectionNode
+end
+
+--- Fixes the ai reverse node rotation for articulated axis vehicles,
+--- if the node is pointing backwards and not forwards.
+function CpAIWorker:getAIReverserNode(superFunc)
+    local spec = self.spec_cpAIWorker
+ --   if not self:getIsCpActive() then return superFunc(self) end
```
Comment (pvaiko) [inline scripts/specializations/CpAIWorker.lua:520]: Is this needed or not?

File: scripts/specializations/CpAIWorker.lua [hunk 8]
```diff
@@ -484,16 +495,86 @@ function CpAIWorker:stopCpDriver()
...
+    end
+    return superFunc(self)
+end
+
+--- Fixes the Direction for the platinum wheel loader, as
+--- their direction is not updated base on the rotation.
+--- So we use the parent node of the arm tool node.
+---@param superFunc any
+function CpAIWorker:getAIDirectionNode(superFunc)
+ --   if not self:getIsCpActive() then return superFunc(self) end
```
Comment (pvaiko) [inline scripts/specializations/CpAIWorker.lua:539]: yes or no?

File: scripts/ai/AIDriveStrategyShovelSiloLoader.lua [hunk 9]
```diff
@@ -0,0 +1,705 @@
+--[[
+This file is part of Courseplay (https://github.com/Courseplay/courseplay)
```
Comment (pvaiko) [inline scripts/ai/AIDriveStrategyShovelSiloLoader.lua:2]: old link

File: scripts/ai/AIDriveStrategyShovelSiloLoader.lua [hunk 10]
```diff
@@ -0,0 +1,705 @@
...
+        self:debug('Pathfinder already active')
+    end
+    return true
+end
+
+function AIDriveStrategyShovelSiloLoader:onPathfindingDoneToStart(path)
+    if path and #path > 2 then
+        self:debug("Found alignment path to the course for the silo.")
+        local alignmentCourse = Course(self.vehicle, CourseGenerator.pointsToXzInPlace(path), true)
+        alignmentCourse:adjustForTowedImplements(2)
```
Comment (pvaiko) [inline scripts/ai/AIDriveStrategyShovelSiloLoader.lua:525]: for this we'll need a better solution in a different PR

File: scripts/ai/Markers.lua [hunk 11]
```diff
@@ -36,23 +43,23 @@ local function setBackMarkerNode(vehicle, measuredBackDistance)
...
-        referenceNode =  AIUtil.getDirectionNode(vehicle)
+        referenceNode = AIUtil.getDirectionNode(vehicle)
         backMarkerOffset = -measuredBackDistance
         CpUtil.debugVehicle(CpDebug.DBG_IMPLEMENTS, vehicle, 'back marker node on measured back distance %.1f', measuredBackDistance)
     elseif reverserNode then
         -- if there is a reverser node, use that, mainly because that most likely will turn with an implement
         -- or with the back component of an articulated vehicle. Just need to find out the distance correctly
-        local dx, _, dz = localToLocal(reverserNode,  AIUtil.getDirectionNode(vehicle), 0, 0, 0)
+        local dx, _, dz = localToLocal(reverserNode, AIUtil.getDirectionNode(vehicle), 0, 0, 0)
         local dBetweenRootAndReverserNode = MathUtil.vector2Length(dx, dz)
```
[the remark this answers is not in the document]
Comment (schwiti6190) [reply scripts/ai/Markers.lua:56]: Fixing this in a separete PR would be better

File: scripts/ai/Markers.lua [hunk 12]
```diff
@@ -36,23 +43,23 @@ local function setBackMarkerNode(vehicle, measuredBackDistance)
...
         backMarkerOffset = dBetweenRootAndReverserNode - vehicle.size.length / 2 - vehicle.size.lengthOffset
         referenceNode = reverserNode
         CpUtil.debugVehicle(CpDebug.DBG_IMPLEMENTS, vehicle, 'Using the %s node for the back marker node %d m from root node (%d m between root and reverser)',
                 debugText, backMarkerOffset, dBetweenRootAndReverserNode)
     else
-        referenceNode =  AIUtil.getDirectionNode(vehicle)
+        referenceNode = AIUtil.getDirectionNode(vehicle)
```
[the remark this answers is not in the document]
Comment (schwiti6190) [reply scripts/ai/Markers.lua:62]: Same as above

File: scripts/pathfinder/PathfinderUtil.lua (hunk 3 above)
Comment (schwiti6190) [reply scripts/pathfinder/PathfinderUtil.lua:925]: I figured it out and plan to easy usage and the boilerplate for using the pathfinder in a later PR

File: scripts/specializations/CpAIWorker.lua (hunk 5 above)
Comment (schwiti6190) [reply scripts/specializations/CpAIWorker.lua:508]: Same as the Reverse questions above, should probably have another look at the reverse code in a separete PR

File: scripts/specializations/CpAIWorker.lua (hunk 6 above)
Comment (schwiti6190) [reply scripts/specializations/CpAIWorker.lua:511]: yes seems so :)

**Review by pvaiko — APPROVED** [review]

**Pull request merged**